README.md
Last modified : 2013-03-20 21:33:48 tkych

Version: 0.1.15 (beta)


CL-SEEK-PROJECT : in Quicklisp, Cliki, GitHub, BitBucket
========================================================

Introduction
------------

CL-SEEK-PROJECT is a search-engine-interface for Common Lisp.
The function SEEK searchs for cl project in Quicklisp, Cliki, Github-Repos, BitBucket-Repos,
then outputs search-results in REPL.


Depends-on
----------

- anaphora
- iterate
- cl-ppcre
- drakma
- flexi-streams
- html-entities


Installation
------------

Quickloadable, just type `(ql:quickload :cl-seek-project)` at REPL.


Examples
--------

```lisp
CL-REPL> (use-package :seek-project)  ;'seek' and '*max-num-description-columns*' are exported symbols.
=> T

CL-REPL> (seek 'Supercalifragilisticexpialidocious)

SEARCH RESULT: "supercalifragilisticexpialidocious"
=> NIL

CL-REPL> (seek "graphviz")

SEARCH-RESULT: "graphviz"

 SEARCH-SPACE: Quicklisp
  com.informatimago.common-lisp.graphviz

 SEARCH-SPACE: Cliki
  cl-graphviz
  cl-dot
  clod
  graph-utils
  Grapher-Server
  s-dot

 SEARCH-SPACE: GitHub-Repos
  cl-graphviz
  cl-dot
  donuts
  protocl
  motion-grammar-kit
  sn-explorer
  compass
  graph-utils
  wo-git-gui
                             ;There is no repo in BitBukect for search-word "graphviz".
                             ; So BitBucket result is not displayed.
=> T

CL-REPL> (seek "graphviz" :description? t :cliki? nil)

SEARCH-RESULT: "graphviz"

 SEARCH-SPACE: Quicklisp
  com.informatimago.common-lisp.graphviz

 SEARCH-SPACE: GitHub-Repos
  cl-graphviz
      Adaptaion of an outdated cl-graphviz by Attila Lendvai to the modern
      version of libgvc.so.6.
  cl-dot
      Common Lisp package for generating GraphViz (dot) files
  donuts
      Graph Drawing DSL (or Graphviz Interface) for Common Lisp.
  protocl
      Native Common Lisp compiler infrastructure for Google's protocol buffers
  motion-grammar-kit
      Formal Language Tools for Robots
  sn-explorer
      A simple web crawler for social networks
  compass
      Recursive, k-width tree clusterer based on the Cardinal Directions
  graph-utils
      graph analysis tools
  wo-git-gui
      WebGUI for git commit graphs, will mainly focus on trying to reduce the
      commit graph into manageable overviews

=> T

CL-REPL> (setf *max-num-description-columns* 60) ;default is 80

=> 60

CL-REPL> (seek "graphviz" :url? t :description? t :cliki? nil :quicklisp? nil)

SEARCH-RESULT: "graphviz"

 SEARCH-SPACE: GitHub-Repos
  cl-graphviz
      https://github.com/necto/cl-graphviz
      Adaptaion of an outdated cl-graphviz by Attila Lendvai
      to the modern version of libgvc.so.6.
  cl-dot
      https://github.com/michaelw/cl-dot
      Common Lisp package for generating GraphViz (dot)
      files
  donuts
      https://github.com/tkych/donuts
      Graph Drawing DSL (or Graphviz Interface) for Common
      Lisp.
  protocl
      https://github.com/scymtym/protocl
      Native Common Lisp compiler infrastructure for
      Google's protocol buffers
  motion-grammar-kit
      https://github.com/golems/motion-grammar-kit
      Formal Language Tools for Robots
  sn-explorer
      https://github.com/GChristensen/sn-explorer
      A simple web crawler for social networks
  compass
      https://github.com/abutcher/compass
      Recursive, k-width tree clusterer based on the
      Cardinal Directions
  graph-utils
      https://github.com/kraison/graph-utils
      graph analysis tools
  wo-git-gui
      https://github.com/woudshoo/wo-git-gui
      WebGUI for git commit graphs, will mainly focus on
      trying to reduce the commit graph into manageable
      overviews

=> T
```    


Referece Manual
---------------

#### [Function] SEEK _search-word_ _&key_ _web?_ _description?_ _url?_ _cliki?_ _github?_ _quicklisp?_ _bitbucket?_

Search for cl project with _search-word_ in Quicklisp, Cliki, GitHub-Repos, BitBucket-Repos.
_search-word_ must be string or symbol (symbol will be coerced to downcase-string).
If _web?_ is NIL, it does not search Cliki, GitHub-Repos and BitBucket-Repos.
If _quicklisp?_ is NIL, it does not search Quicklisp (also _cliki?_, _github?_, _bitbucket?_).
At least one search-space must be specified.
If _description?_ is T, it displays project's description (except for Quicklisp-search).
If _url?_ is T, it displays project's url (except for Quicklisp-search).

- Space in _search-word_:
  If _search-word_ contains #\space, Quicklisp-search is OR-search,
  whereas Cliki,GitHub,BitBucket-search is AND-search.
  e.g. (seek "foo bar"):
       Quicklisp-search    - "foo" OR "bar",
       Cliki,GitHub,BitBucket-search - "foo" AND "bar".

- Max number of search result:
  Quicklisp - unlimited,
  GitHub    - 50,
  Cliki     - 50,
  BitBucket - 50.


#### [Special Variable] \*MAX-NUM-DESCRIPTION-COLUMNS\*

If the length of description-string is bigger then \*max-num-description-columns\*, 
then search-result is inserted newline for easy to see.
Default value is 80.


TODO
----

- Add: limit?, \*max-num-web-search-results\*
- Add: search-space (google-code, etc.)
- Asynchronize: http-request


Author, License, Copyright
--------------------------

* Takaya Ochiai  <#.(reverse "moc.liamg@lper.hcykt")>

* MIT License

* Copyright (C) 2013 Takaya Ochiai


CHANGELOG
Version 0.1.15
2013-03-20
Fix Bug: cliki-search, regex.
       : github-search, remove output's emphasis tags <em>~</em>.

Version 0.1.14
2013-03-10
Fix Bug: github-search, all language, api-v3 -> advanced search
Rename : *description-max-num-cols* -> *max-num-description-columns*

Version 0.1.13
2013-02-26
Add    : pprint for description.
Fix Bug: not displayed BitBucket project whitch have no description.
Rename : seek.lisp + package.lisp -> core.lisp

Version 0.1.12
2013-02-25
Add:    search-space BitBucket-Repos.
Modify: tentative modification for quicklisp-build.

Version 0.1.11
2013-02-22
Rename: Project-Name cl-project-search -> cl-seek-project
      : Thanks Robert Smith for useful suggestion.

Version 0.1.01
2013-02-21
Rename: fn search-project -> seek

Version 0.1.00
2013-01-12
Initial release

Version 0.0.00
2013-01-09
Project Start


seek.lisp
;;;; Last modified : 2013-02-25 21:37:38 tkych

;; cl-seek-project/seek.lisp


;;====================================================================
;; Seek
;;====================================================================

(In-package #:cl-seek-project)


(defparameter *output-description-p* t)
(defparameter *output-url-p* t)

;; !! TODO !! Add: search-space google-code
;; !! TODO !! Pprint: *display-col-max-num-chars*
;; !! TODO !! limit output num:
;;            *display-max-num-projects*, if over limit, y-or-n?
(defun seek (search-word &key (web? t) (description? nil) (url? nil)
                              (cliki? t) (github? t) (bitbucket? t)
                              (quicklisp? t))
  "Search for cl project with SEARCH-WORD in Quicklisp, Cliki, Github-Repos BitBucket-Repos.
SEARCH-WORD must be string or symbol (symbol will be converted to downcase-string).

If WEB? is NIL, it does not search in Cliki, Github-Repos and BitBucket-Repos.
If QUICKLISP? is NIL, it does not search in Quicklisp (also CLIKI?, GITHUB?, BITBUCKET?).
At least one search-space must be specified.

If DESCRIPTION? is T, it displays project's description (except for Quicklisp-search).
If URL? is T, it display project's url (except for Quicklisp-search).

N.B.
 * #\\Space in SEARCH-WORD:
   In case search-word contains #\\Space, Quicklisp-search is OR-search,
   whereas Cliki,GitHub,BitBucket-search is AND-search.

   e.g. (seek \"foo bar\")
        Quicklisp-search for \"foo\" OR \"bar\",
        Cliki,GitHub,BitBucket-search for \"foo\" AND \"bar\".

 * Max number of search result:
    Quicklisp - unlimited
    GitHub    - 100
    Cliki     - 50
    BitBucket - 50
"
  (unless (or (stringp search-word) (symbolp search-word))
    (error "~S is not string or symbol." search-word))
  (unless (or quicklisp?
              (and web? (or cliki? github?)))
    (error "There is no search-space."))
  (let ((found? nil)
        (*output-url-p* url?)
        (word-string
         (write-to-string search-word :case :downcase :escape nil)))
    (format t "~&SEARCH-RESULT: ~S~%" word-string)

    #+quicklisp  ;!? quicklisp is not library, so probably need add this !?
    (when (and quicklisp? (search-quicklisp word-string))
      (setf found? t))

    (when web?
      (let ((drakma:*drakma-default-external-format* :utf-8)
            (*output-description-p* description?))
        (when (and cliki? (search-cliki word-string))
          (setf found? t))
        (when (and github? (search-github word-string))
          (setf found? t))
        (when (and bitbucket? (search-bitbucket word-string))
          (setf found? t))))
    (terpri)
    found?))

;;--------------------------------------------------------------------
;; Quicklisp search
;;--------------------------------------------------------------------
#+quicklisp  ;!? quicklisp is not library, so probably need add this !?
(progn
  (defun search-quicklisp (word-string)
    (let ((found? nil))
      (dolist (wd (ppcre:split " " word-string))
        (awhen (search-ql-systems wd)
          (unless found?
            (format t "~% SEARCH-SPACE: Quicklisp~%")
            (setf found? t))
          (output-ql-results it)))
      found?))

  (defun search-ql-systems (word-string)
    (loop :for system :in (ql-dist:provided-systems t)
          :when (or (search word-string (ql-dist:name system))
                    (search word-string
                            (ql-dist:name (ql-dist:release system))))
          :collect system))

  (defun output-ql-results (systems)
    (dolist (system systems)
      (format t "~&  ~A" (ql-dist:name system))
      ;; (when *output-url-p*
      ;;   (format t "~%      "
      ;;           ))
      ;; (when *output-description-p*
      ;;   (format t "~%      ~A"
      ;;           (ql-dist:short-description system)))
      (terpri)))

  ) ;end of #+quicklisp


;;--------------------------------------------------------------------
;; Cliki search
;;--------------------------------------------------------------------
(defun gen-cliki-query (word-string)
  (format nil "http://www.cliki.net/site/search?query=~A"
          (ppcre:regex-replace-all " " word-string "+")))

(defun search-cliki (word-string)
  (let* ((query   (gen-cliki-query word-string))
         (res     (drakma:http-request query))
         (results (extract-cliki-results res)))
    (when results
      (format t "~% SEARCH-SPACE: Cliki~%")
      (output-cliki-results results)
      (awhen (extract-cliki-rest-page-url res)
        (loop :for q :in it
              :for r := (drakma:http-request q)
              :do (output-cliki-results (extract-cliki-results r))))
      (terpri)
      t)))

(defun extract-cliki-results (res)
  (let ((<li>s (ppcre:all-matches-as-strings "(?s)<li>(.+?)</li>"
                 (ppcre:scan-to-strings
                  "(?s)<ol start=.+?>(.+?)</ol>" res))))
    (when <li>s
      (iter (for <li> :in <li>s)
            (ppcre:register-groups-bind (url title description)
                ("(?s)<li><a href=\"(.+?)\" class=\"internal\">(.+?)</a>\\s?<br\\s?/?>(.+?)</li>"
                 <li>)
              (collect (list title
                             (when *output-url-p*
                               url)
                             (when *output-description-p*
                               (strip (remove-tags description))))))))))

(defun strip (string)
  (string-trim '(#\Space #\Return #\Newline) string))

(defun remove-tags (string)
  (ppcre:regex-replace-all "(<.+?>)" string ""))

(defun output-cliki-results (results)
  (loop :for (title url description) :in results
        :do (format t "~&  ~A" (html-entities:decode-entities title))
            (when *output-url-p*
              (format t "~%      http://www.cliki.net~A" url))
            (when *output-description-p*
              (format t "~%      ~A"
                      (html-entities:decode-entities description)))))

(defparameter *cliki-max-num-of-result-pages* 4) ;total 50

;; extract-cliki-rest-page-url
(defun extract-cliki-rest-page-url (res)
  (let ((urls nil)
        (paginator (ppcre:scan-to-strings
                    "(?s)<div id=\"paginator\">(.+?)</div>" res)))
    (ppcre:do-register-groups (query)
        ("<a href=\"\\\?query=(.+?)\">" paginator)
      (push (gen-cliki-query query) urls))
    (let ((rest-urls (nreverse (rest urls)))) ;first and last is the same.
      (subseq rest-urls
              0 (min *cliki-max-num-of-result-pages*
                     (length rest-urls))))))


;;--------------------------------------------------------------------
;; GitHub search
;;--------------------------------------------------------------------
;; drakma's default url-encoder does not support %encoding.
;; e.g. in url, want "Common Lisp" -> "Common%20Lisp"
;;              but  "Common Lisp" -> error
;;                   "Common+Lisp" -> "Common%2BLisp"
;;                   "Common%20Lisp" -> "Common%2520Lisp"
;; so, (do-urlencode:urlencode search-word) in gen-github-query
;;     (drakma:http-request query :preserve-uri t) in search-github

(defun gen-github-query (search-word)
  (format nil "https://api.github.com/legacy/repos/search/~A~
               ?language=Common%20Lisp"
          (do-urlencode:urlencode search-word)))

;; !! TODO !! max num result is 100. <- github api-v3
;;            How to know there exists results more than 100?
(defun search-github (search-word)
  (let* ((query (gen-github-query search-word))
         (res   (drakma:http-request query :preserve-uri t))
         (jason (yason:parse (flexi-streams:octets-to-string
                              res :external-format :utf-8)))
         (repos (gethash "repositories" jason)))
    (when repos
      (format t "~% SEARCH-SPACE: GitHub-Repos~%")
      (dolist (repo repos)
        (unless (gethash "fork" repo)   ;only master is displayed
          (format t "~&  ~A" (gethash "name" repo))
          (when *output-url-p*
            (format t "~%      https://github.com/~A/~A"
                     (gethash "username" repo) (gethash "name" repo)))
          (awhen (and *output-description-p*
                      (gethash "description" repo))
            (format t "~%      ~A" it))))
      (terpri)
      t)))


;;--------------------------------------------------------------------
;; BitBucket search
;;--------------------------------------------------------------------
;; !!FIX ME: the project does not have 'description' is not displayed.
;;           In fn extract-bitbucket-results,
;;           (ppcre:register-groups-bind (url title description) ...

(defun search-bitbucket (word-string)
  (let* ((query   (gen-bitbucket-query word-string))
         (res     (drakma:http-request query :preserve-uri t))
         (results (extract-bitbucket-results res)))
    (when results
      (format t "~% SEARCH-SPACE: BitBucket-Repos~%")
      (output-bitbucket-results results)
      (awhen (extract-bitbucket-rest-page-url res)
        (loop :for q :in it
              :for r := (drakma:http-request q)
              :do (output-bitbucket-results
                   (extract-bitbucket-results r))))
      (terpri)
      t)))

(defun gen-bitbucket-query (word-string)
  (format nil "https://bitbucket.org/repo/all/relevance?name=~A&language=common+lisp"
          (ppcre:regex-replace-all " " word-string "+")))

(defun extract-bitbucket-results (res)
  (let ((<article>s (ppcre:all-matches-as-strings
                     "(?s)<article class=\"repo-summary\">(.+?)</article>"
                     (ppcre:scan-to-strings
                      "(?s)<section id=\"repo-list\">(.+?)</section>"
                      res))))
    (when <article>s
      (iter (for <article> :in <article>s)
            (ppcre:register-groups-bind (url title description)
                ("(?s)<a class=\"repo-link\" href=\"(.+?)\">.+? / (.+?)</a>.+?<p>(.+?)</p>"
                 <article>)
              (collect (list title
                             (when *output-url-p*
                               url)
                             (when *output-description-p*
                               (strip (remove-tags description))))))))))

(defun output-bitbucket-results (results)
  (loop :for (title url description) :in results
        :do (format t "~&  ~A" (html-entities:decode-entities title))
            (when *output-url-p*
              (format t "~%      https://bitbucket.org~A" url))
            (when *output-description-p*
              (format t "~%      ~A"
                      (html-entities:decode-entities description)))))

(defparameter *bitbucket-max-num-of-result-pages* 4) ;total 50

(defun extract-bitbucket-rest-page-url (res)
  (let ((urls nil)
        (paginator (ppcre:scan-to-strings
                    "(?s)<ol class=\"paginator\">(.+?)</ol>" res)))
    (ppcre:do-register-groups (next-url)
        ("<a href=\"(.+?)\">" paginator)
      (push next-url urls))
    (let ((rest-urls (nreverse (rest urls)))) ;first and last is the same.
      (subseq rest-urls
              0 (min *bitbucket-max-num-of-result-pages*
                     (length rest-urls))))))


;;====================================================================


core.lisp
;;;; Last modified : 2013-03-20 21:48:12 tkych

;; cl-seek-project/core.lisp


;;====================================================================
;; Core
;;====================================================================

(in-package :cl-user)

(defpackage #:cl-seek-project
  (:nicknames #:seek-project)
  (:use :cl :iterate)
  (:import-from #:anaphora #:aif #:awhen #:it)
  (:export #:seek
           #:*max-num-description-columns*
           ;#:*max-num-web-search-results*
           ))

(in-package #:cl-seek-project)


;;--------------------------------------------------------------------
;; Seek
;;--------------------------------------------------------------------
(defparameter *description-print-p* t)
(defparameter *url-print-p* t)

;; !! TODO !! add: search-space google-code
;; !! TODO !! limit output num:
(defun seek (search-word &key (web? t) (description? nil) (url? nil)
                              ;(limit? 50)
                              (cliki? t) (github? t) (bitbucket? t)
                              (quicklisp? t))
  "Search for cl project with SEARCH-WORD in Quicklisp, Cliki, Github-Repos BitBucket-Repos.
SEARCH-WORD must be string or symbol (symbol will be converted to downcase-string).

If WEB? is NIL, it does not search in Cliki, Github-Repos and BitBucket-Repos.
If QUICKLISP? is NIL, it does not search in Quicklisp (also CLIKI?, GITHUB?, BITBUCKET?).
At least one search-space must be specified.

If DESCRIPTION? is T, it displays project's description (except for Quicklisp-search).
If URL? is T, it display project's url (except for Quicklisp-search).

N.B.
 * #\\Space in SEARCH-WORD:
   In case search-word contains #\\Space, Quicklisp-search is OR-search,
   whereas Cliki,GitHub,BitBucket-search is AND-search.

   e.g. (seek \"foo bar\")
        Quicklisp-search for \"foo\" OR \"bar\",
        Cliki,GitHub,BitBucket-search for \"foo\" AND \"bar\".

 * Max number of search result:
    Quicklisp - unlimited
    GitHub    - 50
    Cliki     - 50
    BitBucket - 50
"
  (unless (or (stringp search-word) (symbolp search-word))
    (error "~S is not string or symbol." search-word))
  (unless (or quicklisp?
              (and web? (or cliki? github?)))
    (error "There is no search-space."))
  (let ((found? nil)
        (*url-print-p* url?)
        (word-string
         (write-to-string search-word :case :downcase :escape nil)))
    (format t "~%SEARCH-RESULTS: ~S~%" word-string)

    #+quicklisp   ; for quicklisp build
    (when (and quicklisp? (search-quicklisp word-string))
      (setf found? t))

    (when web?
      (let ((drakma:*drakma-default-external-format* :utf-8)
            ;(*max-num-web-search-results* limit?)
            (*description-print-p* description?))
        (when (and cliki? (search-cliki word-string))
          (setf found? t))
        (when (and github? (search-github word-string))
          (setf found? t))
        (when (and bitbucket? (search-bitbucket word-string))
          (setf found? t))))
    (terpri)
    found?))


;;--------------------------------------------------------------------
;; Quicklisp search
;;--------------------------------------------------------------------
#+quicklisp  ; for quicklisp build
(progn
  (defun search-quicklisp (word-string)
    (let ((found? nil))
      (dolist (wd (ppcre:split " " word-string))
        (awhen (search-ql-systems wd)
          (unless found?
            (format t "~% SEARCH-SPACE: Quicklisp~%")
            (setf found? t))
          (print-ql-results it)))
      found?))

  (defun search-ql-systems (word-string)
    (loop :for system :in (ql-dist:provided-systems t)
          :when (or (search word-string (ql-dist:name system))
                    (search word-string
                            (ql-dist:name (ql-dist:release system))))
          :collect system))

  (defun print-ql-results (systems)
    (dolist (system systems)
      (format t "~&  ~A" (ql-dist:name system))
      ;; (when *url-print-p*
      ;;   (format t "~%      ~A"
      ;;           ))
      ;; (when *description-print-p*
      ;;   (format t "~%      ~A"
      ;;           (ql-dist:short-description system)))
      (terpri)))

  ) ;end of #+quicklisp


;;--------------------------------------------------------------------
;; web search common functions
;;--------------------------------------------------------------------
(defparameter *max-num-web-search-results* 50
  "")

(defparameter *num-results-per-page* 10)

(defun max-num-next-pages ()
  (ceiling (- *max-num-web-search-results* *num-results-per-page*)
           *num-results-per-page*))

(defparameter *max-num-description-columns* 80
  "If the length of description-string is bigger then *max-num-description-columns*,
then description is inserted newline for easy to see.
Default value is 80.")

(defparameter *description-indent-num* 6)

(defun pprint-description (desc)
  (let ((len (length desc))
        (max-nchars (- *max-num-description-columns*
                       *description-indent-num*)))
    (if (<= len max-nchars)
        (format t "~%      ~A" desc)
        (let ((space-index
                (loop :for i :downfrom max-nchars :to 0
                      :until (char= #\Space (char desc i))
                      :finally (return (1+ i)))))
          (if (zerop space-index)
              (progn
                (format t "~%      ~A-" (subseq desc 0 (1- max-nchars)))
                (pprint-description (subseq desc (1- max-nchars))))
              (progn
                (format t "~%      ~A" (subseq desc 0 (1- space-index)))
                (pprint-description (subseq desc space-index))))))))

(defun strip (string)
  (string-trim '(#\Space #\Return #\Newline) string))

(defun remove-tags (string)
  (ppcre:regex-replace-all "(<.+?>)" string ""))

(defun gen-query (format-string word-string)
  (format nil format-string
          (nsubstitute #\+ #\Space word-string :test #'char=)))


;;--------------------------------------------------------------------
;; Cliki search
;;--------------------------------------------------------------------
(defun search-cliki (word-string)
  (let* ((query    (gen-cliki-query word-string))
         (response (drakma:http-request query))
         (results  (extract-cliki-results response)))
    (when results
      (format t "~% SEARCH-SPACE: Cliki~%")
      (print-cliki-results results)
      (awhen (extract-cliki-next-page-url response)
        (loop :for url :in it
              :for res := (drakma:http-request url)
              :do (print-cliki-results (extract-cliki-results res))))
      (terpri)
      t)))

(defun gen-cliki-query (word-string)
  (gen-query "http://www.cliki.net/site/search?query=~A" word-string))

(defun extract-cliki-results (response)
  (let* ((results (ppcre:scan-to-strings
                   "(?s)<ol start=.+?>(.+?)</ol>" response))
         (repos (ppcre:all-matches-as-strings
                 "(?s)<li>(.+?)</li>" results)))
    (when repos
      (iter (for repo :in repos)
            (ppcre:register-groups-bind (url title description)
                ("(?s)<li><a href=\"(.+?)\" class=\"internal\">(.+?)</a>\\s?<br\\s?/?>(.+?)</li>"
                 repo)
              (collect (list title
                             (when *url-print-p* url)
                             (when *description-print-p*
                               (strip (remove-tags description))))))))))

(defun print-cliki-results (results)
  (loop :for (title url desc) :in results :do
     (format t "~&  ~A" (html-entities:decode-entities title))
     (when *url-print-p*
       (format t "~%      http://www.cliki.net~A" url))
     (when *description-print-p*
       (pprint-description
        (html-entities:decode-entities desc)))))

(defun extract-cliki-next-page-url (response)
  (let ((urls nil)
        (paginator (ppcre:scan-to-strings
                    "(?s)<div id=\"paginator\">(.+?)</div>" response)))
    (ppcre:do-register-groups (query)
        ("<a href=\"\\\?query=(.+?)\">" paginator)
      (push (gen-cliki-query query) urls))
    (let ((rest-urls (nreverse (rest urls)))) ;first and last is the same.
      (subseq rest-urls
              0 (min (max-num-next-pages) (length rest-urls))))))


;;--------------------------------------------------------------------
;; GitHub search
;;--------------------------------------------------------------------
;; 2013-03-10
;; Changed: github-search api-v3 legacy search -> advanced search.
;;          For github api-v3, search option '?language=' does not work??.

;; drakma's default url-encoder does not support %encoding.
;; e.g. in url, want "Common Lisp" -> "Common%20Lisp"
;;              but  "Common Lisp" -> error
;;                   "Common+Lisp" -> "Common%2BLisp"
;;                   "Common%20Lisp" -> "Common%2520Lisp"
;; so, (do-urlencode:urlencode search-word) in gen-github-query
;;     (drakma:http-request query :preserve-uri t) in search-github
;; (defun gen-github-query (search-word)
;;   (format nil "https://api.github.com/legacy/repos/search/~A~
;;                ?language=Common%20Lisp"
;;           (do-urlencode:urlencode search-word)))

;; !! TODO !! yason: hash-table -> alist
;; (defun search-github (search-word)
;;   (let* ((query (gen-github-query search-word))
;;          (res   (drakma:http-request query :preserve-uri t))
;;          (jason (yason:parse (flexi-streams:octets-to-string
;;                               res :external-format :utf-8)))
;;          (repos (gethash "repositories" jason)))
;;     (when repos
;;       (format t "~% SEARCH-SPACE: GitHub-Repos~%")
;;       (dolist (repo repos)
;;         (unless (gethash "fork" repo)   ;only master is displayed
;;           (format t "~&  ~A" (gethash "name" repo))
;;           (when *output-url-p*
;;             (format t "~%      https://github.com/~A/~A"
;;                      (gethash "username" repo) (gethash "name" repo)))
;;           (awhen (and *output-description-p*
;;                       (gethash "description" repo))
;;             (unless (string= "" it)
;;               (pprint-description it)))))
;;       (terpri)
;;       t)))


(defun search-github (word-string)
  (let* ((query    (gen-github-query word-string))
         (response (drakma:http-request query :preserve-uri t))
         (results  (extract-github-results response)))
    (when results
      (format t "~% SEARCH-SPACE: GitHub-Repos~%")
      (print-github-results results)
      (awhen (extract-github-next-page-url response)
        (loop :for url :in it
              :for res := (drakma:http-request url :preserve-uri t)
              :do (print-github-results
                   (extract-github-results res))))
      (terpri)
      t)))

(defun gen-github-query (word-string)
  (gen-query "https://github.com/search?q=~A~
              &type=Repositories&ref=advsearch&l=Common+Lisp"
             word-string))

(defun extract-github-results (response)
  (let* ((results (ppcre:scan-to-strings
                   "(?s)<ul class=\"repolist js-repo-list\">(.+<!-- /.body -->)"
                   response))
         (repos (ppcre:all-matches-as-strings
                 "(?s)<h3>(.+?)</p>"
                 results)))
    (when repos
      (iter (for repo :in repos)
            (ppcre:register-groups-bind (url title)
                ("(?s)<h3>.+?<a href=\"/(.+?)\">.+?/(.+?)</a>" repo)
              (collect
                  (list (ppcre:regex-replace-all "</?em>" title "")
                        (when *url-print-p* url)
                        (when *description-print-p*
                          (ppcre:register-groups-bind
                              (description)
                              ("(?s)<p class=\"description\">(.+?)</p>" repo)
                            (strip (remove-tags description)))))))))))

(defun print-github-results (results)
  (loop :for (title url desc) :in results :do
     (format t "~&  ~A" (html-entities:decode-entities title))
     (when *url-print-p*
       (format t "~%      https://github.com/~A" url))
     (when (and *description-print-p* desc)
       (pprint-description
        (html-entities:decode-entities desc)))))

(defun extract-github-next-page-url (response)
  (let ((urls nil)
        (pagination
         (ppcre:scan-to-strings
          "(?s)<div class=\"pagination\"(.+?)</div>" response)))
    (ppcre:do-register-groups (next-url)
        ("<a href=\"(.+?)\"( (class|rel).+?)?>" pagination)
      (push (format nil "https://github.com~A"
                    (html-entities:decode-entities next-url))
            urls))
    (let ((rest-urls (nreverse (rest urls)))) ;first and last is the same.
      (subseq rest-urls
              0 (min (max-num-next-pages) (length rest-urls))))))


;;--------------------------------------------------------------------
;; BitBucket search
;;--------------------------------------------------------------------
(defun search-bitbucket (word-string)
  (let* ((query    (gen-bitbucket-query word-string))
         (response (drakma:http-request query :preserve-uri t))
         (results  (extract-bitbucket-results response)))
    (when results
      (format t "~% SEARCH-SPACE: BitBucket-Repos~%")
      (print-bitbucket-results results)
      (awhen (extract-bitbucket-next-page-url response)
        (loop :for url :in it
              :for res := (drakma:http-request url)
              :do (print-bitbucket-results
                   (extract-bitbucket-results res))))
      (terpri)
      t)))

(defun gen-bitbucket-query (word-string)
  (gen-query "https://bitbucket.org/repo/all/relevance?name=~A~
              &language=common+lisp"
             word-string))

(defun extract-bitbucket-results (response)
  (let* ((results (ppcre:scan-to-strings
                   "(?s)<section id=\"repo-list\">(.+?)</section>"
                   response))
         (repos (ppcre:all-matches-as-strings
                 "(?s)<article class=\"repo-summary\">(.+?)</article>"
                 results)))
    (when repos
      (iter (for repo :in repos)
            (ppcre:register-groups-bind (url title)
                ("(?s)<a class=\"repo-link\" href=\"(.+?)\">.+? / (.+?)</a>"
                 repo)
              (collect (list title
                             (when *url-print-p* url)
                             (when *description-print-p*
                               (ppcre:register-groups-bind (description)
                                   ("(?s)<p>(.+?)</p>" repo)
                                 (strip (remove-tags description)))))))))))

(defun print-bitbucket-results (results)
  (loop :for (title url desc) :in results :do
     (format t "~&  ~A" (html-entities:decode-entities title))
     (when *url-print-p*
       (format t "~%      https://bitbucket.org~A" url))
     (when (and *description-print-p* desc)
       (pprint-description
        (html-entities:decode-entities desc)))))

(defun extract-bitbucket-next-page-url (response)
  (let ((urls nil)
        (paginator (ppcre:scan-to-strings
                    "(?s)<ol class=\"paginator\">(.+?)</ol>" response)))
    (ppcre:do-register-groups (next-url)
        ("<a href=\"(.+?)\">" paginator)
      (push next-url urls))
    (let ((rest-urls (nreverse (rest urls)))) ;first and last is the same.
      (subseq rest-urls
              0 (min (max-num-next-pages) (length rest-urls))))))


;;====================================================================
